dbms/src/Interpreters/Aggregator.h

```diff
@@ -890,6 +885,47 @@ struct AggregatedDataVariants : private boost::noncopyable
 using AggregatedDataVariantsPtr = std::shared_ptr<AggregatedDataVariants>;
 using ManyAggregatedDataVariants = std::vector<AggregatedDataVariantsPtr>;
 
+/// Combines aggregation states together, turns them into blocks, and outputs.
+class MergingBuckets
+{
+public:
+    /** The input is a set of non-empty sets of partially aggregated data,
+      *  which are all either single-level, or are two-level.
+      */
+    MergingBuckets(const Aggregator & aggregator_, const ManyAggregatedDataVariants & data_, bool final_, size_t concurrency_);
+
+    Block getHeader() const;
+
+    Block getData(size_t concurrency_index);
+
+    size_t getConcurrency() const { return concurrency; }
+
+private:
+    Block getDataForSingleLevel();
+
+    Block getDataForTwoLevel(size_t concurrency_index);
+
+    void doLevelMerge(Int32 bucket_num, size_t concurrency_index);
+
+private:
+    const LoggerPtr log;
+    const Aggregator & aggregator;
+    ManyAggregatedDataVariants data;
+    bool final;
```

Comment:
You are right, but I don't want to rename it in this pr..
Since final is used in many places such `Aggregator`, `AggregatingBlockInputStream`, and `MergingAggregatedMemoryEfficientBlockInputStream` and etc, it might be better to do it with another pr..